File: frontend/types/dataset.ts
export type InferredRole =
  | "numeric"
  | "categorical"
  | "datetime"
  | "identifier"
  | "boolean"
  | "text"
  | "unknown";

export interface ColumnTypeInfo {
  name: string;
  pandas_dtype: string;
  inferred_role: InferredRole;
  unique_count: number;
  unique_ratio: number;
  role_hints: string[];
}

export interface DatasetOverview {
  dataset_id: string;
  filename: string;
  n_rows: number;
  n_columns: number;
  column_types: ColumnTypeInfo[];
  preview_rows: Record<string, unknown>[];
  uploaded_at: string;
}

export interface ColumnQualityReport {
  column: string;
  dtype: string;
  missing_count: number;
  missing_rate: number;
  duplicate_value_count: number;
  is_constant: boolean;
  outlier_count: number | null;
  outlier_method: string | null;
  zero_ratio: number | null;
  negative_ratio: number | null;
  skewness: number | null;
  kurtosis: number | null;
  suggested_transformation: string | null;
  reason: string | null;
}

export interface DatasetQualityReport {
  dataset_id: string;
  n_rows: number;
  n_columns: number;
  duplicate_row_count: number;
  duplicate_row_rate: number;
  constant_columns: string[];
  potential_id_columns: string[];
  potential_time_columns: string[];
  potential_categorical_columns: string[];
  columns: ColumnQualityReport[];
}

export type DatasetType = "panel" | "time_series" | "cross_sectional" | "unknown";

export interface StructureDetectionResult {
  dataset_type: DatasetType;
  entity_column: string | null;
  time_column: string | null;
  is_balanced_panel: boolean | null;
  entity_count: number | null;
  time_period_count: number | null;
  explanation: string;
}

export interface DatasetProfileResponse {
  dataset_id: string;
  quality: DatasetQualityReport;
  structure: StructureDetectionResult;
}

export interface ApiErrorPayload {
  error: {
    code: string;
    message: string;
    details: Record<string, unknown>;
  };
}

File: backend/app/schemas/dataset.py
"""Pydantic schemas for dataset upload, preview, and profiling."""
from __future__ import annotations

from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


class ColumnTypeInfo(BaseModel):
    name: str
    pandas_dtype: str
    inferred_role: Literal[
        "numeric",
        "categorical",
        "datetime",
        "identifier",
        "boolean",
        "text",
        "unknown",
    ]
    unique_count: int
    unique_ratio: float
    role_hints: list[str] = Field(
        default_factory=list,
        description="e.g. Potential Outcome, Possible Entity ID, Possible Time Variable",
    )


class DatasetSummary(BaseModel):
    dataset_id: str
    filename: str
    n_rows: int
    n_columns: int
    columns: list[str]
    uploaded_at: datetime


class DatasetOverviewResponse(BaseModel):
    dataset_id: str
    filename: str
    n_rows: int
    n_columns: int
    column_types: list[ColumnTypeInfo]
    preview_rows: list[dict[str, Any]]
    uploaded_at: datetime


class ColumnQualityReport(BaseModel):
    column: str
    dtype: str
    missing_count: int
    missing_rate: float
    duplicate_value_count: int = 0
    is_constant: bool
    outlier_count: int | None = None
    outlier_method: str | None = None
    zero_ratio: float | None = None
    negative_ratio: float | None = None
    skewness: float | None = None
    kurtosis: float | None = None
    suggested_transformation: str | None = None
    reason: str | None = None


class DatasetQualityReport(BaseModel):
    dataset_id: str
    n_rows: int
    n_columns: int
    duplicate_row_count: int
    duplicate_row_rate: float
    constant_columns: list[str]
    potential_id_columns: list[str]
    potential_time_columns: list[str]
    potential_categorical_columns: list[str]
    columns: list[ColumnQualityReport]


class StructureDetectionResult(BaseModel):
    dataset_type: Literal["panel", "time_series", "cross_sectional", "unknown"]
    entity_column: str | None = None
    time_column: str | None = None
    is_balanced_panel: bool | None = None
    entity_count: int | None = None
    time_period_count: int | None = None
    explanation: str


class DatasetProfileResponse(BaseModel):
    dataset_id: str
    quality: DatasetQualityReport
    structure: StructureDetectionResult
